Fix pdm bincounting

nabilbrice opened the pull request:
<!--
Thanks for creating a pull request! Please ensure visiting our [contributor's guide](https://github.com/StingraySoftware/stingray/blob/main/CONTRIBUTING.md)

Please do not submit empty sections and remove them as per your requirements
-->

#### Relevant Issue(s)/PR(s)
Fixes #871.

#### Provide an overview of the implemented solution or the fix and elaborate on the modifications.
I added a test to demonstrate the bug, which fails with the old implementation using ```np.unique``` with ```return_values=True``` to count the number of instances in each bin. The new implementation is a one line change, which calls ```np.histogram``` instead, re-using the ```bins``` information. The ```bins``` are offset in the call to ```np.histogram``` to include the integer values of ```bin_idx```.

#### Is there a new dependency introduced by your contribution? If so, please specify.
No.

#### Any other comments?
I encountered this bug when trying to use ```fold_events``` on sparely sampled Swift data.

matteobachetti commented on stingray/pulse/pulsar.py:
```diff
@@ -328,7 +327,7 @@ def fold_events(times, *frequency_derivatives, **opts):
         # I need the variance uncorrected for the number of data points in each
         # bin, so I need to find that first, and then multiply
-        _, bincounts = np.unique(bin_idx, return_counts=True)
+        bincounts, _ = np.histogram(bin_idx, bins=bins * 10 + 0.5)
```
Why are the bins rescaled like this? In particular, I do not understand the factor 10. Adding actual numerical checks in the test below would probably help

nabilbrice replied:
Thanks for the review, Matteo. The rescaling was actually an error on my part - it was specific to the problem I was solving (with ```10``` bins). I've now changed it to construct the bin edges directly with ```np.arange(0.5, nbin + 1.5)```, which should be the correct bin-edges to bin indices running from ```[1, ..., nbin]```. I've added a few comments above too. Let me know if it's clear!

The changes are in [ea955d7](https://github.com/StingraySoftware/stingray/pull/872/commits/ea955d7955b116e8a3106675649a624befea4252).

matteobachetti reviewed (changes requested):
@nabilbrice thanks a lot for catching the bug, and for fixing it!

I don't understand how you decide the bins though, would you mind adding a comment to the code explaining the rescale? Also, an actual numerical test, providing a comparison between the obtained `profile`, `prof_err` etc. with the expected values, would be very helpful.

An additional small request: a changelog entry for towncrier, as described here: https://docs.stingray.science/en/stable/contributing.html#updating-and-maintaining-the-changelog

nabilbrice commented:
Hi @matteobachetti, thanks for the quick review! I have added the ```872.bugfix.rst``` for ```towncrier``` now in [32a8694](https://github.com/StingraySoftware/stingray/pull/872/commits/32a869434d81bfb711429e724b4a8feab3209ddc).

I'm not sure what could be done for the numerical test on ```profile``` and ```prof_err``` that would specifically catch this situation. The error that comes up when using the old implementation of ```fold_events``` in [c74c616](https://github.com/StingraySoftware/stingray/pull/872/commits/c74c61603231f772a471328264c94827c4c6d8ed) is to do with a rank mismatch between ```raw_profile``` and ```bincounts```, which stops the multiplication from executing. The test I added doesn't really have changes to the expected outputs with the new implementation, instead it won't throw an error.

Would it be better to change one of the previous tests instead to set it up so that it would break because a bin occupancy is ```0```?

matteobachetti commented:
@nabilbrice Now that the issue is fixed, the current test you added should basically always run. If you could just add a condition on the outputs of this (e.g. the folded profile should be equal to _something_ ), it would be more than sufficient. If you can do the same using existing tests, great too. I just realized that the test we have only tests the prof_err array, and not the profile, which is not ideal

nabilbrice commented:
> @nabilbrice Now that the issue is fixed, the current test you added should basically always run. If you could just add a condition on the outputs of this (e.g. the folded profile should be equal to _something_ ), it would be more than sufficient. If you can do the same using existing tests, great too. I just realized that the test we have only tests the prof_err array, and not the profile, which is not ideal

@matteobachetti, since the new test wasn't actually testing anything concrete, I've combined it with the previous ```test_pulse_profile_pdm``` by changing the frequency in ```fold_events``` from ```1``` to ```0.237```. This results in a failing test with the old implementation (rank mismatch error) but is otherwise not a well motivated value for the frequency. The changes can be found in [689837a](https://github.com/StingraySoftware/stingray/pull/872/commits/689837a6e1b3113cda731b2dc4794b8873a008d2).

While I'm still poking around this section of the code, I'd be happy to open a different issue for the tests on the ```pulse_profile_pdm```. Currently, there doesn't seem to be a test on the values in the ```profile``` because it is randomly generated. It might be good to set up a consistent set of mock data to test against.

matteobachetti reviewed (approved):
@nabilbrice Thanks for your catch and for fixing it! We'll take care of improving the test in the future.
